Add support for remote temp file location for queued chunk reading

cmorbitzer opened the pull request:
**Draft** for preliminary comments and review, still needing tests and a couple safety checks.

Mark the following tasks as done:

* [x] Checked the codebase to ensure that your feature doesn't already exist.
* [x] Checked the pull requests to ensure that another person hasn't already submitted the feature or fix.
* [ ] Adjusted the Documentation.
* [x] Added tests to ensure against regression.

### Description of the Change

This adds support for a new configuration key `remote_temp_disk` that, when not null, will store and retrieve spreadsheet chunks for queued reading at the specified filesystem disk in addition to the configured `temp_path` on the local server.

### Why Should This Be Added?

Currently queued chunk reading is impossible on a multi-server setup, because it cannot be guaranteed that the spreadsheet chunk, a temporary file, will be uploaded and stored on the same server as the queue worker. In this case the `ReadChunk` job throws `InvalidArgumentException: File "/tmp/xxx" does not exist.`

### Benefits

<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks

<!-- What are the possible side-effects or negative impacts (e.g. breaking changes) of the code change? -->

### Verification Process

<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

-->

### Applicable Issues

This is similar to, but not the same as, #1839, #1697, and #1827.

cmorbitzer committed:
Add support for remote temp file location for queued chunk reading

patrickbrouwers commented:
Hey @cmorbitzer thanks for the PR. I was thinking about doing something similar. I'm not really a fan of having multiple ways of dealing with temp files, so I'll probably merge that into 1; moving the normal temp file handling to the filesystem as well (default disk). Main goal is to not have a performance penalty when not using a clustered queue setup - as getting the file from S3 does slow down the process a bit.

patrickbrouwers commented on config/excel.php:
```diff
@@ -86,6 +86,22 @@
+        | can be available to both servers.
+        |
+        */
+        'remote_temp_disk'       => null,
```
Lets make this a global setting (so not have difference for import/export) called `temporary_file_location`, It would be nice if this can either be a full path (`sys_get_temp_dir`) or either a disk name (`null` is default disk, `s3`, etc).

patrickbrouwers commented on src/Jobs/ReadChunk.php:
```diff
@@ -86,6 +87,10 @@ public function handle()
         $this->reader->setReadEmptyCells(false);
 
+        if ($this->sheetImport instanceof ShouldQueue && ($remoteTempDisk = config('excel.imports.remote_temp_disk')) !== null) {
+            app(FilePathHelper::class)->copyToTempFile(basename($this->file), $this->file, $remoteTempDisk);
```
Inject `FilePathHelper` in `handle()`

patrickbrouwers commented on src/Jobs/ReadChunk.php:
```diff
@@ -86,6 +87,10 @@ public function handle()
         $this->reader->setReadDataOnly(true);
         $this->reader->setReadEmptyCells(false);
 
+        if ($this->sheetImport instanceof ShouldQueue && ($remoteTempDisk = config('excel.imports.remote_temp_disk')) !== null) {
```
No if-statement, make this default behaviour. Handle logic in copyToTempFile

patrickbrouwers commented on src/Reader.php:
```diff
@@ -284,6 +284,10 @@ private function getReader($import, $filePath, string $readerType = null, string
 
         $this->currentFile = $this->filePathHelper->getRealPath($filePath, $disk);
 
+        if ($import instanceof ShouldQueue && ($remoteTempDisk = config('excel.imports.remote_temp_disk')) !== null) {
```
Make default behaviour, merge with line above. getRealPath should store temp file on a disk + return the tmp file.

cmorbitzer commented:
@patrickbrouwers Thanks for the feedback! It's well received. I'll make the changes today for another review, as we have a release blocked by this.

patrickbrouwers commented:
Thanks @cmorbitzer !

cmorbitzer committed:
Use temp file location globally

cmorbitzer commented:
@patrickbrouwers I addressed your comments as best as possible. Because PhpSpreadsheet requires files to be on the local filesystem, there was no way to consolidate both settings into one. There always has to be a local temp path even if we also need a separate (potentially remote) temp location.

The FilePathHelper has been substantially refactored to handle just about everything regarding temp files. This DRYs a lot of the existing code.

The remote_temp_disk will only be used for queued imports/exports because otherwise we know that the same server will be handling the entire job. You had said you didn't want to have two ways of handling temp files, but it seemed unnecessary to me to always be uploading and downloading temp files from a potentially remote location when we know it's just not necessary and would be a waste of resources. By having the FilePathHelper handle most everything, I hoped to make this as easy to understand and maintainable as possible. So in this case, since your desire for maintainability and performance were in conflict, I chose to side with performance.

This also includes some various minor refactors I noticed along the way. Feel free to revert those if I;m lacking institutional knowledge as to why it was the way it was.

I did not test this in any way yet, or give it a second look myself. I'm going to step away for a few hours before beginning that process. I wanted to commit this in case you or anyone else had comments in the meantime.

patrickbrouwers commented:
hey @cmorbitzer it looks fine to me. If you can make sure everything works in both queued as non-queued situations, than I'm happy to polish the last details myself after merging tomorrow.

cmorbitzer committed:
Minor changes

cmorbitzer committed:
Merge branch '3.1' into feature/remote-temp-file

cmorbitzer committed:
Existing tests pass

cmorbitzer committed:
Style

cmorbitzer committed:
Style??

cmorbitzer committed:
Add tests

cmorbitzer committed:
Appease styleci
